src/graphics/Screen.cpp

```diff
@@ -1400,21 +1415,28 @@ int Screen::handleTextMessage(const meshtastic_MeshPacket *packet)
             // Incoming message
             devicestate.has_rx_text_message = true; // Needed to include the message frame
             hasUnreadMessage = true;                // Enables mail icon in the header
-            setFrames(FOCUS_PRESERVE);              // Refresh frame list without switching view
 
-            // Only wake/force display if the configuration allows it
-            if (shouldWakeOnReceivedMessage()) {
-                setOn(true);    // Wake up the screen first
-                forceDisplay(); // Forces screen redraw
+            // Always update frame list to include new message, but defer UI updates if virtual keyboard is active
+            if (NotificationRenderer::current_notification_type != notificationTypeEnum::text_input) {
+                setFrames(FOCUS_PRESERVE); // Refresh frame list without switching view
+
+                // Only wake/force display if the configuration allows it
+                if (shouldWakeOnReceivedMessage()) {
+                    setOn(true);    // Wake up the screen first
+                    forceDisplay(); // Forces screen redraw
+                }
+            } else {
+                // Virtual keyboard is active - just mark that frames need regeneration when keyboard closes
+                // The devicestate and hasUnreadMessage are already set above, so message will appear later
+                LOG_INFO("Virtual keyboard active - deferring frame list update for new message");
```

Comment:
This should probably be LOG_DEBUG level, rather than LOG_INFO, since the audience of the message is developers, rather than users.